### ISSUE TITLE
Federation: Limit resolver parent types to contain only fields from @key and @requires directives

### ISSUE THREAD

**Issue by jakeblaxon** [dotansimha/graphql-code-generator#4089]
**Is your feature request related to a problem? Please describe.**
With federation, we now have the ability to determine exactly which parent fields are guaranteed to be available to a resolver. In theory, these are only the fields specified in the `@key` directive for the parent type and the `@requires` directive for the field. (Note: this only applies to the fields of extended types).

**Describe the solution you'd like**
Using this information, we should generate unique parent types for each field's resolver such that they only contain the subset of fields specified by these directives. This would give developers better type safety, since they'll be able to know when they're trying to use a parent field that's not part of the `@key` or `@requires` directives.

**Describe alternatives you've considered**
We could instead just make all the other parent fields that are not in these directives optional, which would force developers to handle their null cases. But that would obscure the nullability information provided for those fields in the schema, and it would still incorrectly indicate that those fields may not be null in some cases.

**Additional context**
For example, consider the following federated type:
```graphql
extend type Book @key(fields: "isbn") {
    isbn: String! @external
    title: String! @external
    numberOfPages: Int! @external
    author: Author @requires(fields: "title")
    publisher: Publisher @requires(fields: "numberOfPages")
}

type Author {
    name: String!
}

type Publisher {
    name: String!
}
```

The parent type for the `Book.author` resolver that currently gets generated contains all fields found in the `Book` type, but since `numberOfPages` isn't included in the `@key` or `@requires` directives, this field should be excluded from the parent type. The `publisher` field should also be excluded from the parent type.

Likewise for the `Book.publisher` resolver, but the parent type shouldn't include `title` and `author` in this case.

**Comment by ardatan** [comment]
@jakeblaxon PRs are welcome!

**Comment by dotansimha** [comment]
Fixed in `v1.17.4`
